lib/charms/operator_libs_linux/v1/systemd.py

```diff
@@ -66,165 +70,219 @@
-        cmd.append(service_name)
-    if now is not None:
-        cmd.append("--now")
-    if quiet is not None:
-        cmd.append("--quiet")
-    if sub_cmd != "is-active":
-        logger.debug("Attempting to {} '{}' with command {}.".format(cmd, service_name, cmd))
-    else:
-        logger.debug("Checking if '{}' is active".format(service_name))
-
-    proc = subprocess.Popen(cmd, **_popen_kwargs())
-    last_line = ""
-    for line in iter(proc.stdout.readline, ""):
-        last_line = line
-        logger.debug(line)
-
-    proc.wait()
-
-    if proc.returncode < 1:
-        return True
-
-    # If we are just checking whether a service is running, return True/False, rather
-    # than raising an error.
-    if sub_cmd == "is-active" and proc.returncode == 3:  # Code returned when service not active.
-        return False
-
-    if sub_cmd == "is-failed":
-        return False
-
-    raise SystemdError(
-        "Could not {}{}: systemd output: {}".format(
-            sub_cmd, " {}".format(service_name) if service_name else "", last_line
+    cmd = ["systemctl", *args]
+    logger.debug(f"Executing command: {cmd}")
+    try:
+        proc = subprocess.run(
+            cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            text=True,
+            bufsize=1,
+            encoding="utf-8",
+            check=check,
+        )
+        logger.debug(
+            f"Command {cmd} exit code: {proc.returncode}. systemctl output:\n{proc.stdout}"
+        )
+        return proc.returncode
+    except subprocess.CalledProcessError as e:
+        raise SystemdError(
+            f"Command {cmd} failed with returncode {e.returncode}. systemctl output:\n{e.stdout}"
         )
-    )
 
 
 def service_running(service_name: str) -> bool:
-    """Determine whether a system service is running.
+    """Report whether a system service is running.
 
     Args:
-        service_name: the name of the service to check
+        service_name: The name of the service to check.
+
+    Return:
+        True if service is running/active; False if not.
     """
-    return _systemctl("is-active", service_name, quiet=True)
+    # If returncode is 0, this means that is service is active.
+    return _systemctl("--quiet", "is-active", service_name) == 0
 
 
 def service_failed(service_name: str) -> bool:
-    """Determine whether a system service has failed.
+    """Report whether a system service has failed.
 
     Args:
-        service_name: the name of the service to check
+        service_name: The name of the service to check.
+
+    Returns:
+        True if service is marked as failed; False if not.
     """
-    return _systemctl("is-failed", service_name, quiet=True)
+    # If returncode is 0, this means that the service has failed.
+    return _systemctl("--quiet", "is-failed", service_name) == 0
 
 
-def service_start(service_name: str) -> bool:
+def service_start(*args: str) -> bool:
```

Comment:
This should probably be named `*service_names` rather than the generic `*args` -- when you're just using `*args` or `**kwargs` as a pass-through and you don't know the meaning, use the generic terms, but when the argument has specific meaning (the services names in this case) it's best to use a proper name.

Reply:
Hmm... normally I would agree with you that `*<parameter_name>` should capture the meaning of the positional arguments being passed to the method, however, I don't find that `*service_names` is a good fit here. Reason saying is that `service_start` | `service_stop` | `service_restart` can (and should) accept more than just service names as position arguments.

For example, the I have a use case where I need to pass the `--no-block` flag to `systemctl start ...` because the service I am working with is `Type=forking` and the parent process can take potentially take ~15 minutes to return control to systemd. Long story... but I feel that if we change `*args` -> `*service_names`, charm authors consuming this lib may not realize that they are allowed to pass flags to `systemctl` as well as service names. Therefore, I think that `*args` is a better parameter name in this situation since the `service_*` methods now allow you to pass generic arguments to `systemctl` via `subprocess.run([...])`.

Reply:
Ah, I missed that nuance -- makes sense!